Carnap-Server/Settings/RuntimeDefs.hs

```diff
@@ -0,0 +1,35 @@
+-- Defines the runtime settings. You probably don't want to import this file,
+-- use Settings.Runtime instead which reexports everything from here.
+module Settings.RuntimeDefs where
+
+import           ClassyPrelude
+import           Database.Persist.TH
+import qualified Data.Aeson as A
+
+data RTSetType =
+      TyDisableGoogleReg
+    -- ^ disable new registrations of accounts with the Google login back end
+    | TyInstanceAdminEmail
+    -- ^ the email for the administrator of this Carnap instance
+    deriving (Show, Read, Eq, Enum)
+
+data RTSetting =
+      DisableGoogleReg Bool
+    | InstanceAdminEmail Text
+    deriving (Show, Eq)
+
+rtSettingType :: RTSetting -> RTSetType
+rtSettingType (DisableGoogleReg _) = TyDisableGoogleReg
+rtSettingType (InstanceAdminEmail _) = TyInstanceAdminEmail
+
+parseRtSetting :: RTSetType -> ByteString -> Maybe RTSetting
```

Comment:
Question: Why parse and serialize rather than deriving a new persist field?

Reply:
So, the process that led to this design was this:

* Lookups should be indexed (need two columns, name and value)
* It must be type safe on the Haskell side and not take too much boilerplate

Because of this, I ended up with this design. I think you could equally put the type in the name field and the data sum type in the data field, but they could be mismatched, leading to somewhat unfortunate API on the haskell side (imo). 

The other complexity driving the haskell side is that I needed to be able to list, as strings, the variants of the settings type to produce the settings thing automatically. 

I am still unsure if it's possible to do this more cleverly by leaning harder on the type system and using a type class or something. I feel like this pattern should be possible to do in a cleaner way but I'm blanking on how. If you've got ideas, I'd appreciate them.

Reply:
OK. I get need for two columns. The thing I'm blanking on a bit is why the second column should be a bytestring rather than just including the RTSetting itself, via `derivePersistField`.

Reply:
You'd have to write a fallible function to convert from it and check the id matches the type, which would work but it would be effectively the same code. Another reason for doing it this way is that you can reuse the parsing function to also handle the user input from the browser, which I think you'd need some hack with `read` to do the other way.

Reply:
Got it. OK, I've read a little more carefully now and I think I understand how this works better. This seems like a reasonable way of setting things up.